includes/course-video/blocks/class-sensei-course-video-blocks-extension-abstract.php

```diff
@@ -0,0 +1,119 @@
+<?php
+/**
+ * File containing the Sensei_Course_Video_Blocks_Extension_Abstract class.
+ *
+ * @package sensei-lms
+ * @since 4.0.0
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Provides common logic for video block extensions to implement functionality for video course progression
+ *
+ * @since 4.0.0
+ */
+abstract class Sensei_Course_Video_Blocks_Extension_Abstract {
+	/**
+	 * Instance of class.
+	 *
+	 * @var self
+	 */
+	private static $instance;
+
+	/**
+	 * Manages Video course-related settings.
+	 *
+	 * @var Sensei_Course_Video_Settings
+	 */
+	private $settings;
+
+	/**
+	 * Returns an instance of the class.
+	 *
+	 * @param Sensei_Course_Video_Settings $settings
+	 *
+	 * @return static
+	 */
+	public static function instance( Sensei_Course_Video_Settings $settings ) {
+		if ( static::$instance ) {
+			return static::$instance;
+		}
+
+		static::$instance = new static( $settings );
+		return static::$instance;
+	}
+
+	/**
+	 * Initialize the class and hooks.
+	 *
+	 * @param Sensei_Course_Video_Settings $settings
+	 */
+	public static function init( Sensei_Course_Video_Settings $settings ) {
+		static::instance( $settings )->init_hooks();
+	}
+
+	/**
+	 * Sensei_Youtube_Extension constructor.
+	 *
+	 * @param Sensei_Course_Video_Settings $settings
+	 */
+	private function __construct( Sensei_Course_Video_Settings $settings ) {
+		$this->settings = $settings;
+	}
+
+	/**
+	 * Initialize hooks.
+	 */
+	public function init_hooks() {
+		add_filter( $this->get_filter_name(), [ $this, 'wrap_block' ], 10, $this->get_filter_params_number() );
+	}
+
+	/**
+	 * Returns name of the filter to be used for the video block extension
+	 *
+	 * @return string
+	 */
+	abstract protected function get_filter_name();
+
+	/**
+	 * Returns number of parameters that the filter accepts.
+	 *
+	 * @return int
+	 */
+	abstract protected function get_filter_params_number();
+
+	/**
+	 * Returns wrapped video block.
+	 *
+	 * @param array $params
+	 *
+	 * @return mixed
+	 */
+	abstract protected function wrap_block( ...$params );
```

Comment:
I think it is much worse then explicit method with parameters.
But I didn't think up how to handle variable number of params for different filters.